app/code/Magento/QuoteGraphQl/Model/Resolver/AddProductsToCart.php

```diff
@@ -0,0 +1,90 @@
+<?php
+/**
+ * Copyright © Magento, Inc. All rights reserved.
+ * See COPYING.txt for license details.
+ */
+declare(strict_types=1);
+
+namespace Magento\QuoteGraphQl\Model\Resolver;
+
+use Magento\Framework\GraphQl\Config\Element\Field;
+use Magento\Framework\GraphQl\Exception\GraphQlInputException;
+use Magento\Framework\GraphQl\Query\ResolverInterface;
+use Magento\Framework\GraphQl\Schema\Type\ResolveInfo;
+use Magento\Quote\Model\Cart\Data\AddProductsToCartOutput;
+use Magento\Quote\Model\Cart\Data\CartItemFactory;
+use Magento\QuoteGraphQl\Model\Cart\GetCartForUser;
+use Magento\Quote\Model\Cart\Data\Error;
+
+/**
+ * Add simple products to cart GraphQl resolver
+ * {@inheritdoc}
+ */
+class AddProductsToCart implements ResolverInterface
+{
+    /**
+     * @var GetCartForUser
+     */
+    private $getCartForUser;
+
+    /**
+     * @var \Magento\Quote\Model\Cart\AddProductsToCart
+     */
+    private $addProductsToCart;
+
+    /**
+     * @param GetCartForUser $getCartForUser
+     * @param \Magento\Quote\Model\Cart\AddProductsToCart $addProductsToCart
+     */
+    public function __construct(
+        GetCartForUser $getCartForUser,
+        \Magento\Quote\Model\Cart\AddProductsToCart $addProductsToCart
+    ) {
+        $this->getCartForUser = $getCartForUser;
+        $this->addProductsToCart = $addProductsToCart;
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public function resolve(Field $field, $context, ResolveInfo $info, array $value = null, array $args = null)
+    {
+        if (empty($args['cart_id'])) {
+            throw new GraphQlInputException(__('Required parameter "cart_id" is missing'));
+        }
+        if (empty($args['cart_items']) || !is_array($args['cart_items'])
+        ) {
+            throw new GraphQlInputException(__('Required parameter "cart_items" is missing'));
+        }
+        $maskedCartId = $args['cart_id'];
+        $cartItemsData = $args['cart_items'];
+        $storeId = (int)$context->getExtensionAttributes()->getStore()->getId();
+
+        // Shopping Cart validation
+        $this->getCartForUser->execute($maskedCartId, $context->getUserId(), $storeId);
+
+        $cartItems = [];
+        foreach ($cartItemsData as $cartItemData) {
+            $cartItems[] = (new CartItemFactory())->create($cartItemData);
+        }
+
+        /** @var AddProductsToCartOutput $addProductsToCartOutput */
+        $addProductsToCartOutput = $this->addProductsToCart->execute($maskedCartId, $cartItems);
+
+        return [
```

Comment:
With mentioned exception we should be able to return both - cart data and errors.

Reply:
@rogyar :
Below is a list of few of the use cases that are failing.  Please take a look.
Case 1: add product to cart with a qty> the stock available 
Actual result: 
```
"errors": {
      "debugMessage": "Expected a value of type \"CheckoutUserInputErrorCodes\" but received: (empty string)","message": "Internal server error",
```
Case 2 : Add product with qty=0 and a valid sku, 

> Expected: cart is empty and error message to enter a quantity >0
> Actual: The product is still added to the cart with qty =1

Case 3: Add disabled product to cart
> Expected: Cart is empty and message that "product is not available"
> Actual result : 
```
exception: {
  "errors": [
    {
      "debugMessage": "Expected a value of type \"CheckoutUserInputErrorCodes\" but received: (empty string)",
      "message": "Internal server error",
Now, if we omit "code" from userInputError{} and run the same query, below exception is returned
exception: ""errors": [
    {
      "debugMessage": "User Error: expected iterable, but did not find one for field CheckoutUserInputError.path.","
```
Case 4 : Same result as above when you try to add a product with "Out of Stock" status.